### PR TITLE
Use source.LANG/text.LANG as language map fallbacks for ST 4

### PR DESCRIPTION
**Pull request by jfcherng**
Probably resolves https://github.com/facelessuser/sublime-markdown-popups/issues/102 @rchl @rwols 

---

Answer to concerns in #102:

- This feature is activated if and only if `sublime.find_syntax_by_scope`, which is introduced in the very first ST 4 dev build, is available.
- Since this is a fallback, users can still set the correct dialect in the language map if that's what they want.
- Language map is still preferred and there is no need to maintain two version of `mdpopups` or language map.

### CODE REVIEW

**Comment by facelessuser** [comment]
Thanks for doing this. I will evaluate and let you know my thoughts. I appreciate the approach to making this a fallback. I feel doing this as a "Hail Mary pass" when we are desperate to find a suitable language due to missing entries in the mapping is a fair approach.

File: st3/mdpopups/st_code_highlight.py [hunk 1]
```diff
@@ -250,6 +257,12 @@ def set_view(self, src, lang, plugin_map):
                         break
             if loaded:
                 break
+        if not loaded:
+            # Use "source.LANG" and "text.LANG" as fallbacks if possible
```
Comment (facelessuser) [inline st3/mdpopups/st_code_highlight.py:261]: Can you add `fallbacks` to https://github.com/facelessuser/sublime-markdown-popups/blob/master/docs/src/dictionary/en-custom.txt?

Comment (jfcherng) [reply st3/mdpopups/st_code_highlight.py:261]: ok

File: st3/mdpopups/st_code_highlight.py [hunk 2]
```diff
@@ -237,6 +237,13 @@ def set_view(self, src, lang, plugin_map):
             user_v = user_map.get(key, (tuple(), tuple()))
             if lang in (tuple(user_v[0]) + plugin_v[0] + v[0]):
                 for l in (tuple(user_v[1]) + plugin_v[1] + v[1]):
+                    if l.startswith('scope:'):
```
Comment (facelessuser) [inline st3/mdpopups/st_code_highlight.py:240]: So, I take this part is meant to allow users to describe their mapping in terms of scope?

Comment (jfcherng) [reply st3/mdpopups/st_code_highlight.py:240]: yes.

Comment (facelessuser) [reply st3/mdpopups/st_code_highlight.py:240]: Okay, the original description didn't highlight this. I had thought this was just a fallback. So the user is doing this?

````
```scope:source.js
content...
```
````

Comment (facelessuser) [reply st3/mdpopups/st_code_highlight.py:240]: Or is this used as a way in the mapping to link syntax. Forgive me, I haven't touched this area of the code a bit, and I haven't looked at the full context.

Comment (jfcherng) [reply st3/mdpopups/st_code_highlight.py:240]: > Okay, the original description didn't highlight this. I had thought this was just a fallback. So the user is doing this?
> 
> ```
> ```scope:source.js
> content...
> ```
> ```

Technically users can do this as it's a side effect of this PR. But that's not my purpose.

Comment (facelessuser) [reply st3/mdpopups/st_code_highlight.py:240]: Ah, okay, that makes sense. I'm not sure colons would make it through fenced code Markdown syntax anyways, though I honestly don't recall.

Comment (jfcherng) [reply st3/mdpopups/st_code_highlight.py:240]: > Okay, the original description didn't highlight this. I had thought this was just a fallback. So the user is doing this?
> 
> ```
> ```scope:source.js
> content...
> ```
> ```

That sets `lang` to `scope:source.js` and the fallback will match nothing as there is no scope `source.scope:source.js` or `text.scope:source.js`.

**Comment by facelessuser** [comment]
The last thing we need is to update documentation on this. I'm thinking that overall this looks good.

**Comment by facelessuser** [comment]
@gir-bot lgtm

**Comment by facelessuser** [comment]
I'll go ahead and merge this and play around with it locally. I don't see any big concerns, so it is quite likely to make it into the next release.

**Pull request merged**